### PR TITLE
Add targets for the extract processing stage

### PR DESCRIPTION
**Pull request by Jennit07**
Opening a draft PR so that we can see files which have changed on this branch. This is still in the testing phase, however, I think it is at a good point to review some changes and gradually bring code back into main-R. The top-level script provides a skeleton for bringing functions together and processing extracts prior to creating the episode file. This so far includes lookups/costs, social care 'all' files, and extracts/tests which are all needed prior to processing the episode file. The top-level function does take a while to process extracts into a list and I've still to test the social care (year-specific) data in this. 

There are improvements which can be made to this so any ideas are welcomed. One thought I had was during the testing: if there was something breaking the processing of a function, this would lock up my R session and not return the data as a list. An Idea @Moohan had was to add an optional parameter for checking a file exists on disk and if it does read in that file to the list, if it does not then run the processing code. OR similarly, if the list returns data for a specific extract and it exists in the list then do nothing, if the data is not in the list then run the processing code.

### CODE REVIEW

File: inst/WORDLIST [hunk 1]
```diff
@@ -51,7 +52,9 @@ gpprac
 gz
 hbpraccode
 hscdiip
+keydate
 lca
+mnodification
```
Comment (Moohan) [inline inst/WORDLIST:57]:
```suggestion
```

File: R/process_extract_acute.R [hunk 2]
```diff
@@ -40,7 +40,7 @@ process_extract_acute <- function(data, year, write_to_disk = TRUE) {
     ) %>%
     # Recode GP practice into 5 digit number
     # We assume that if it starts with a letter it's an English practice and so recode to 99995.
-    dplyr::mutate(gpprac = convert_eng_gpprac_to_dummy(.data$gpprac)) %>%
+    dplyr::mutate(gpprac = convert_eng_gpprac_to_dummy(data$gpprac)) %>%
```
Comment (Moohan) [inline R/process_extract_acute.R:43]:
```suggestion
    dplyr::mutate(gpprac = convert_eng_gpprac_to_dummy(.data$gpprac)) %>%
```

**Review by Moohan — CHANGES_REQUESTED** [review]
I only spotted one small error which I'm pretty sure won't be causing any of the issues you're getting.

I think this PR is ready to have some of the code split off into smaller PRs so they can be checked and merged separately though!

File: Production_scripts/Phase_2_production/phase_2_production.R [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+# Create a list of the years to run
+years <- convert_year_to_fyyear(as.character(2017)) %>%
+  purrr::set_names()
+
+
+------------------------------------------------------
+# Process Lookups
+------------------------------------------------------
+# Lookups/costs
+process_lookups <- run_process_lookups()
```
[the remark this answers is not in the document]
Comment (Jennit07) [reply Production_scripts/Phase_2_production/phase_2_production.R:27]: Test for github overview. Please ignore

**Comment by Jennit07** [comment]
Ready for review to bring this branch back into mainR @Moohan

File: R/read_costs_gp_ooh.R
Comment (Moohan) [inline R/read_costs_gp_ooh.R:1]: Split out to another PR

File: Rmarkdown/hc_costs.Rmd
Comment (Moohan) [inline Rmarkdown/hc_costs.Rmd:1]: Split out to another PR

**Review by Moohan — COMMENTED** [review]
First pass. Mostly some comments around splitting code out to it's own PR where the changes are unrelated.

File: R/produce_tests_pis.R
[the remark this answers is not in the document]
Comment (Jennit07) [reply R/produce_tests_pis.R:1]: Already split out - will remove from this branch

File: Rmarkdown/ch_costs.Rmd
[the remark this answers is not in the document]
Comment (Jennit07) [reply Rmarkdown/ch_costs.Rmd:1]: Already split out - will remove from this branch

File: Rmarkdown/gp_ooh_costs.Rmd
[the remark this answers is not in the document]
Comment (Jennit07) [reply Rmarkdown/gp_ooh_costs.Rmd:1]: Already split out - will remove from this branch

File: Rmarkdown/hc_costs.Rmd
Comment (Jennit07) [reply Rmarkdown/hc_costs.Rmd:1]: Already split out - will remove from this branch

File: R/read_costs_gp_ooh.R
Comment (Jennit07) [reply R/read_costs_gp_ooh.R:1]: Already split out - will remove from this branch

File: R/read_costs_gp_ooh.R [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    overwrite = TRUE
+  )
+
+  # Read in data---------------------------------------
+
+  # Costs spreadsheet
+  gp_ooh_data <- readxl::read_xlsx(paste0(
+    get_slf_dir(),
+    "/Costs/OOH_Costs.xlsx"
+  ))
```
[the remark this answers is not in the document]
Comment (Jennit07) [reply R/read_costs_gp_ooh.R:20]: Already split out - will remove from this branch. This is already updated to use this on Main so no need to update the function on MainR

File: R/run_data_extracts.R [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  process_extracts <- list(
+    "acute" = process_extract_acute(read_extract_acute(year), year, write_to_disk = write_to_disk),
+    "ae" = process_extract_ae(read_extract_ae(year), year, write_to_disk = write_to_disk),
+    "mental_health" = process_extract_mental_health(read_extract_mental_health(year), year, write_to_disk = write_to_disk),
+    "maternity" = process_extract_maternity(read_extract_maternity(year), year, write_to_disk = write_to_disk),
+    "nrs_deaths" = process_extract_nrs_deaths(read_extract_nrs_deaths(year), year, write_to_disk = write_to_disk),
+    "outpatients" = process_extract_outpatients(read_extract_outpatients(year), year, write_to_disk = write_to_disk),
+    "pis" = process_extract_prescribing(read_extract_prescribing(year), year, write_to_disk = write_to_disk) # ,
+    # "ltc" = process_lookup_ltc(read_lookup_ltc(), year, write_to_disk = write_to_disk)
+    # "ooh" = process_extract_ooh(read_extract_ooh(year), year, write_to_disk = write_to_disk)
```
[the remark this answers is not in the document]
Comment (Jennit07) [reply R/run_data_extracts.R:24]: I think this should be deleted as this was using the 'list' format - forgot to remove some scripts on this branch

File: R/run_process_social_care.R [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    "sds" = process_sc_all_sds(read_sc_all_sds(sc_dvprod_connection),
+      sc_demographics = sc_demographics,
+      write_to_disk = write_to_disk
+    )
+  )
+
+  return(process_social_care)
+}
+
+# "care_home" = process_sc_all_care_home(write_to_disk = write_to_disk),
```
[the remark this answers is not in the document]
Comment (Jennit07) [reply R/run_process_social_care.R:35]: This script should be deleted. This was following the 'list' processing idea before using targets.

File: R/run_data_extracts.R (hunk 5 above)
[the remark this answers is not in the document]
Comment (Jennit07) [reply R/run_data_extracts.R]: Deleted this

File: R/run_process_social_care.R (hunk 6 above)
[the remark this answers is not in the document]
Comment (Jennit07) [reply R/run_process_social_care.R]: deleted this

File: _targets/meta/meta
[the remark this answers is not in the document]
Comment (Jennit07) [reply _targets/meta/meta:1]: Change made, not sure if i did this correctly though, could you please confirm? Thanks

File: _targets.R [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+  # format = "parquet",
+  resources = tar_resources(
+    # parquet = tar_resources_parquet(compression = "zstd"),
+    future = tar_resources_future(plan = future::multisession)
+  ),
+  error = "continue",
+  # default option
+  storage = "main",
+  # default option
+  memory = "persistant"
```
Comment (Moohan) [inline _targets.R:18]:
```suggestion
  memory = "persistent"
```

File: _targets.R [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+  # format = "parquet",
+  resources = tar_resources(
+    # parquet = tar_resources_parquet(compression = "zstd"),
+    future = tar_resources_future(plan = future::multisession)
+  ),
+  error = "continue",
+  # default option
+  storage = "main",
+  # default option
+  memory = "persistent"
```
Comment (Moohan) [inline _targets.R:18]:
```suggestion
  storage = "main",   # default option
  memory = "persistent"   # default option
```

File: R/process_extract_gp_ooh.R [hunk 9]
```diff
@@ -82,6 +82,10 @@ process_extract_gp_ooh <- function(year, data_list, write_to_disk = TRUE) {
     dplyr::mutate(
       record_keydate1 = as.Date(.data$record_keydate1),
       record_keydate2 = as.Date(.data$record_keydate2)
+    ) %>%
+    # change attendance_status into numeric
+    mutate(
+      attendance_status = as.double(attendance_status)
```
Comment (Moohan) [inline R/process_extract_gp_ooh.R:88]: I don't think this will work properly as it's already a factor from [`read_extract_ooh...`](https://github.com/Public-Health-Scotland/source-linkage-files/blob/64c51db19a051ce173620af372942a7b17b926e4/R/read_extract_ooh_consultations.R#L16) so when you convert it (back?) to a numeric, it will become `1` or `2` but actually we want it to match the `attendance_status` [values for outpatients](https://www.ndc.scot.nhs.uk/Dictionary-A-Z/Definitions/index.asp?Search=A&ID=96&Title=Attendance%20Status) which are 1 (seen) or 8 (not seen).

Comment (Moohan) [reply R/process_extract_gp_ooh.R:88]: We should probably change the GP OoH `attendance_status` (`"Consultation Recorded"`) and the outpatient `attendance_status` ("Clinic Attendance (Status) Code") to both be a factor with something like:
`factor(<ooh_var>, levels = c("Y", "N"), labels = c("Patient Seen", "DNA"))` and
`factor(<outpatient_var>, levels = c(1, 5, 8), labels = c("Patient Seen", "CNW", "DNA"))`.

But this would also involve tweaking some other code which uses the 1 / 8 values, so out of scope for this PR. For now, I think it's sufficient to alter the GP OoH values this just needs to be done at the right place.

File: DESCRIPTION [hunk 10]
```diff
@@ -58,6 +58,8 @@ Suggests:
     covr (>= 3.6.1),
     roxygen2 (>= 7.2.3),
     spelling (>= 2.2),
+    tarchetypes,
+    targets,
```
Comment (Moohan) [inline DESCRIPTION:62]: Make packages the current latest versions.
```suggestion
    tarchetypes (>= 0.7.5),
    targets (>= 0.14.3),
```

**Review by Moohan — APPROVED** [review]
Other than 1718 Care Home which has been highlighted in #634 everything is working from start to finish for the targets pipeline.

This is tested using 1718 and 1920 ran in parallel I used an 8CPU 64GB session and it worked fine. The main point I'd note is that all_sds runs very slow compared to everything else so would be worth inspecting!
